src/librustc_mir/build/scope.rs

```diff
@@ -828,6 +879,27 @@ fn build_diverge_scope<'a, 'gcx, 'tcx>(tcx: TyCtxt<'a, 'gcx, 'tcx>,
         };
     }
 
+    let repatching;
+    if scope.diverge_path.terminator_emitted {
+        // we need to re-patch the terminator so that it now points to
+        // any drops that were inserted above.
+        cfg.block_data_mut(scope.diverge_path.block).terminator = None;
+        scope.diverge_path.repatch_count += 1;
+        repatching = format!("repatch {}", scope.diverge_path.repatch_count);
```

Comment:
mmm there's two control flow paths that initialize `repatching`, and the two closure expressions would have distinct types. I've looked into using `&Fn() -> String` but then the borrowck says the closures are temporaries that do not live long enough. Are you intending that I go all the way to `Box<Fn() -> String>`? (Which does work, and I guess is slightly less intensive than formatting a string unconditionally, but its not as clear to be a significant win...)